rcli/typing.py

```diff
@@ -0,0 +1,665 @@
+
+            def __new__(cls, __value, *args, **kwargs):
+                # type: (Any, *Any, **Any) -> _T
+                """Return __value cast to _T.
+
+                Any additional arguments are passed as-is to the constructor.
+
+                Args:
+                    __value: A value that can be converted to type _T.
+                    args: Any additional positional arguments passed to the
+                        constructor.
+                    kwargs: Any additional keyword arguments passed to the
+                        constructor.
+                """
+                ret = type_(__value, *args, **kwargs)
+                expected = keyfunc(__value)
+                if bound.start and expected < bound.start:
+                    if keyfunc is not identity:
+                        raise ValueError(
+                            'The value of {}({}) [{}] is below the minimum '
+                            'allowed value of {}.'.format(
+                                keyfunc_name, repr(__value), repr(expected),
+                                bound.start))
+                    raise ValueError(
+                        'The value {} is below the minimum allowed value '
+                        'of {}.'.format(repr(__value), bound.start))
+                if bound.stop and expected > bound.stop:
+                    if keyfunc is not identity:
+                        raise ValueError(
+                            'The value of {}({}) [{}] is above the maximum'
+                            ' allowed value of {}.'.format(
+                                keyfunc_name, repr(__value), repr(expected),
+                                bound.stop))
+                    raise ValueError(
+                        'The value {} is above the maximum allowed value '
+                        'of {}.'.format(repr(__value), bound.stop))
+                return ret
+
+        if keyfunc is not cls._default:
+            _BoundedSubclass.__class_repr__ = '{}.{}[{}, {}, {}]'.format(
+                cls.__module__, cls.__name__, cls._get_fullname(type_),
+                cls._get_bound_repr(bound), keyfunc_name)
+        else:
+            _BoundedSubclass.__class_repr__ = '{}.{}[{}, {}]'.format(
+                cls.__module__, cls.__name__, cls._get_fullname(type_),
+                cls._get_bound_repr(bound))
+        return _BoundedSubclass
+
+    def _get_args(cls, args):
+        # type: (tuple) -> Tuple[Any, slice, Callable]
+        """Return the parameters necessary to check type boundaries.
+
+        Args:
+            args: A tuple with two or three parameters: a type, a slice
+                representing the minimum and maximum lengths allowed for values
+                of that type and, optionally, a function to use on values
+                before comparing against the bounds.
+
+        Returns:
+            A tuple with three parameters: a type, a slice, and a function to
```

Comment:
Do we mean to say "elements" here, or do we really mean "parameters"?...    I guess it's context dependent but I think in _this_ context elements is more correct?